### PR TITLE
[NoQA] Add unit testing guide and example tests

### PR DESCRIPTION
**Pull request by OlimpiaZurek**
### Explanation of Change

This PR adds a UI testing Guide with best practices  and basic example tests for components

### Fixed Issues

$ https://github.com/Expensify/App/issues/52771
PROPOSAL:

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

// TODO: These must be filled out, or the issue title must include "[No QA]."

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: tests/ui/components/CheckboxWithLabel.tsx
Comment (mountiny) [inline tests/ui/components/CheckboxWithLabel.tsx:26]:
```suggestion
        renderCheckboxWithLabel();

```

File: tests/README.md
Comment (mountiny) [inline tests/README.md:16]:
```suggestion
- Familiarity with the React Native Testing Library [RNTL](https://callstack.github.io/react-native-testing-library/).
```

File: tests/README.md
Comment (mountiny) [inline tests/README.md:7]: Not sure if we need to have this in here as its quite generic stuff people should know, but also not feeling strongly to remove it

File: tests/README.md
Comment (mountiny) [inline tests/README.md:67]:
```suggestion
```

File: tests/README.md
Comment (mountiny) [inline tests/README.md:111]:
```suggestion
- Non-Functional Changes: E.g styling updates, github workflow changes, translation updates or refactors without behavioral changes.
```

File: tests/README.md
Comment (mountiny) [inline tests/README.md:113]: Not sure about this one, while I can see what you aim for here, I think we should try to create some tests for this cases too and abstract the bug fix into a test whenever possible, even if its an old code

File: tests/README.md
Comment (mountiny) [inline tests/README.md:3]: In general, I think we should move this either later section of this document or to README in the tests/ui

File: tests/README.md
Comment (OlimpiaZurek) [reply tests/README.md]: I agree, I removed that part

File: tests/README.md
Comment (OlimpiaZurek) [reply tests/README.md]: I moved to the tests/ui

File: tests/ui/README.md
Comment (tgolen) [inline tests/ui/README.md:94]: Could you please add more guidance for `act()`? It has always confused me

File: tests/ui/README.md
Comment (tgolen) [inline tests/ui/README.md:118]:
```suggestion
- Not cleaning up between tests: React components often leave behind side effects. Make sure to clean up between tests to ensure they are isolated and avoid conflicts.

- Having tightly coupled tests where one test depends on the actions from a previous test. This can cause tests to fail unexpectedly and it makes tests hard to maintain and debug. Each test should be self-sufficient.
```

File: tests/ui/README.md
Comment (tgolen) [inline tests/ui/README.md:84]:
```suggestion
- Mock external dependencies like network calls or Onyx data. If the amount of onyx data is large, place it in an external JSON file with the same name as the test file and import the data.
```

File: tests/ui/README.md
Comment (tgolen) [inline tests/ui/README.md:87]: I think some more explanation could be added here. What would those attributes be tested for and how should they be tested?

**Review by tgolen — CHANGES_REQUESTED** [review]
I love this. The examples you added are great!

File: tests/ui/README.md
Comment (OlimpiaZurek) [reply tests/ui/README.md]: Added a short explanation and clarified when manual usage is needed

File: tests/ui/README.md
Comment (OlimpiaZurek) [reply tests/ui/README.md]: Added more details and example.

File: tests/ui/README.md
Comment (tgolen) [inline tests/ui/README.md:101]:
```suggestion
- Don’t ignore warnings: Fix `act()` or other common warnings to ensure reliability.  The `act()` function ensures that any updates to the state or props of a component (including rendering, firing events, or manual state updates) are completed before the test proceeds. The RNTL automatically wraps most rendering and interaction utilities (`render`, `fireEvent`, `waitFor`, etc.) in `act()` to manage these updates. However, there are some situations where manual use of `act()` might be necessary, particularly when dealing with asynchronous logic, animations or timers(`setTimeout`, `setInterval` etc). You can find more detailed explanation [here](https://callstack.github.io/react-native-testing-library/docs/advanced/understanding-act).
```

**Review by tgolen — DISMISSED** [review]
Looks great! I just spotted these two small fixes.

File: tests/ui/README.md
Comment (mountiny) [inline tests/ui/README.md:101]:
```suggestion
- Don’t ignore warnings: Fix `act()` or other common warnings to ensure reliability. The `act()`function ensures that any updates to the state or props of a component (including rendering, firing events, or manual state updates) are completed before the test proceeds. The RNTL automatically wraps most rendering and interaction utilities (`render`, `fireEvent`, `waitFor`, etc.) in `act()` to manage these updates. However, there are some situations where manual use of `act()` might be necessary, particularly when dealing with asynchronous logic, animations or timers (`setTimeOut`, `setInterval` etc). You can find more detailed explanation [here](https://callstack.github.io/react-native-testing-library/docs/advanced/understanding-act).
```

**Review by mountiny — DISMISSED** [review]
small comment

**Review by mountiny — APPROVED** [review]
All our comments addressed so I will not hold this back and merge. @puneetlath feel free to review even after this is merged

**Pull request merged**
